app/model/commands/AddAssetCommand.scala

```diff
@@ -0,0 +1,64 @@
+package model.commands
+
+import java.util.Date
+
+import com.gu.atom.data.PreviewDataStore
+import CommandExceptions._
+import com.gu.atom.publish.PreviewAtomPublisher
+import com.gu.contentatom.thrift.{ContentAtomEvent, EventType}
+import com.gu.contentatom.thrift.atom.media.Asset
+import util.atom.MediaAtomImplicits
+import model.ThriftUtil
+
+import scala.util.{Failure, Success}
+
+case class AddAssetCommand(atomId: String,
+                           videoUri: String,
+                           version: Option[Long],
+                           mimeType: Option[String])
+                          (implicit previewDataStore: PreviewDataStore,
+                           previewPublisher: PreviewAtomPublisher)
+    extends Command
+    with MediaAtomImplicits {
+
+  type T = Unit
+
+  def process(): Unit = {
+    previewDataStore.getAtom(atomId) match {
+      case Some(atom) =>
+        val mediaAtom = atom.tdata
+        val currentAssets: Seq[Asset] = mediaAtom.assets
+
+        val resolvedVersion = version.getOrElse(currentAssets.foldLeft(1L){(acc, asset) => if (asset.version >= acc) asset.version + 1 else acc})
+
+        if (currentAssets.exists(asset => asset.version == resolvedVersion && asset.mimeType == mimeType)) {
+          AssetVersionConflict
+        }
+
+        val newAsset = ThriftUtil.parseAsset(videoUri, mimeType, resolvedVersion) match {
+          case Left(_) => AssetParseFailed
+          case Right(a) => a
+        }
```

Comment:
@currysoup Since you like folds so much, we could rewrite this as:
`val newAsset = ThriftUtil.parseAsset(videoUri, mimeType, resolvedVersion).fold(err => AssetParseFailed, identity)`